spacy_experimental/span_finder/span_finder_component.py

```diff
@@ -0,0 +1,322 @@
+                            doc.spans[self.candidates_key].append(doc[start : end + 1])
+                        elif self.max_length > 0 and span_length > self.max_length:
+                            break
+
+    def update(
+        self,
+        examples: Iterable[Example],
+        *,
+        drop: float = 0.0,
+        sgd: Optional[Optimizer] = None,
+        losses: Optional[Dict[str, float]] = None,
+    ) -> Dict[str, float]:
+        """Learn from a batch of documents and gold-standard information,
+        updating the pipe's model. Delegates to predict and get_loss.
+        examples (Iterable[Example]): A batch of Example objects.
+        drop (float): The dropout rate.
+        sgd (Optional[thinc.api.Optimizer]): The optimizer.
+        losses (Optional[Dict[str, float]]): Optional record of the loss during training.
+            Updated using the component name as the key.
+        RETURNS (Dict[str, float]): The updated losses dictionary.
+        """
+        if losses is None:
+            losses = {}
+        losses.setdefault(self.name, 0.0)
+        predicted = [eg.predicted for eg in examples]
+        set_dropout_rate(self.model, drop)
+        scores, backprop_scores = self.model.begin_update(predicted)
+        loss, d_scores = self.get_loss(examples, scores)
+        backprop_scores(d_scores)
+        if sgd is not None:
+            self.finish_update(sgd)
+        losses[self.name] += loss
+        return losses
+
+    def get_loss(self, examples, scores) -> Tuple[float, float]:
+        """Find the loss and gradient of loss for the batch of documents and
+        their predicted scores.
+        examples (Iterable[Examples]): The batch of examples.
+        scores: Scores representing the model's predictions.
+        RETURNS (Tuple[float, float]): The loss and the gradient.
+        """
+        references = [eg.reference for eg in examples]
+        reference_results = self.model.ops.asarray(
+            self._get_reference(references), dtype=float32
+        )
+        d_scores = scores - reference_results
+        loss = float((d_scores**2).sum())
+        return loss, d_scores
+
+    def _get_reference(self, docs) -> Floats2d:
+        """Create a reference list of token probabilities for calculating loss"""
+        reference_results = []
+        for doc in docs:
+            start_indices = set()
+            end_indices = set()
+
+            if self.reference_key in doc.spans:
+                for span in doc.spans[self.reference_key]:
+                    start_indices.add(span.start)
+                    end_indices.add(span.end - 1)
+
+            for token in doc:
+                reference_results.append(
+                    (
+                        1 if token.i in start_indices else 0,
+                        1 if token.i in end_indices else 0,
+                    )
+                )
+
+        return reference_results
```

Comment:
I should have caught this earlier, but there's one more major update needed here:

This needs to be modified to handle cases where the predicted and reference tokenization differ. Instead of extracting the reference indices from the reference doc only, you need to align the annotation between the predicted and reference docs and handle cases where the annotation can't be aligned.

Given that you have binary labels on tokens, I think it would look a lot like the loss in the `senter` component. Because you're converting from underlying span annotation, you can't use `Example.get_aligned` directly, so I'm not sure there's a way around a custom alignment method here that does the span+token alignment for this particular case.

You can handle missing values by using an existing loss function from thinc like `SequenceCategoricalCrossentropy`.

Reply:
I've created a new method `_get_aligned_scores` that validates the alignment between the `predicted` and the `reference` doc and checks for three cases:

1. Tokens are aligned (best case :tada:)
2. Predicted has less tokens than the Reference (`alignment.y2x`)
3. Predicted has more tokens than the Reference (`alignment.x2y`)

It's using the predicted scores and the spans in the reference to create two new lists of scores that are used to calculate the loss. The changes can be found here [18dbe4d](https://github.com/explosion/spacy-experimental/pull/10/commits/18dbe4dd77f38c9fb0d5677e78155046166d1c5d).
There are still some errors when iterating through the alignment properties.